### PR TITLE
updated main.nf to work with merge_bam process

### PR DESCRIPTION
**Pull request by sb43**
@ewels 
Please refer issue [#71 
 implementing bam_merge step for multiple split bam's of a given sample ]( https://github.com/nf-core/methylseq/issues/71)
I have implemented additional step which will merge multiple bams for a sample after alignment.
e.g., SampleA : split_0_R1_fastq.gz, split_0_R2_fastq.gz
SampleA: split_1_R1_fastq.gz, split_1_R2_fastq.gz
SampleB: split_0_R1_fastq.gz, split_0_R2_fastq.gz
SampleB: split_1_R1_fastq.gz, split_1_R2_fastq.gz

Let me know your suggestions.

### CODE REVIEW

**Comment by ewels** [comment]
Many thanks for your contribution @sb43!

It looks like there are a few things that need tidying up before this can be merged. Mostly, it's that you've done your own workarounds to the single-sample bug (#66). I fixed this earlier and have now release v1.3 with those changes, so you'll need to revert your changes and pull in the updates before this pull request can be merged. Let me know once this is done and I'll test and review the changes.

To confirm - the function of this new addition is to merge all input files into a single sample? For example, running on a set of multiplexed fastq files from a single biological sample?

Thanks,

Phil

**Comment by sb43** [comment]
@ewels  you are right,  merge_bams step is to deal with multiplexed samples or large samples where user can split fastq files into chunks and merge back bismark aligned bam's

File: main.nf [hunk 1]
```diff
@@ -544,9 +576,7 @@ if( params.aligner == 'bismark' ){
             """
         }
     }
-
-    ch_bismark_align_log_for_bismark_report
-     .join(ch_bismark_dedup_log_for_bismark_report)
+     ch_bismark_dedup_log_for_bismark_report
```
Comment (ewels) [inline main.nf]: This change will break functionality for regular users. This step is required to synchronise the logs for each sample from each step. Without it, random alignment logs from each sample will be mixed with random log sets from other samples.

File: main.nf [hunk 2]
```diff
@@ -560,8 +590,8 @@ if( params.aligner == 'bismark' ){
         publishDir "${params.outdir}/bismark_reports", mode: 'copy'
 
         input:
-        set val(name), file(align_log), file(dedup_log), file(splitting_report), file(mbias) from ch_bismark_logs_for_bismark_report
-
+        set val(name), file(dedup_log), file(splitting_report), file(mbias) from ch_bismark_logs_for_bismark_report
+        file(align_log) from ch_bismark_align_log_for_bismark_report.collect()
```
Comment (ewels) [inline main.nf]: As above, this will break functionality for existing usage.

File: main.nf (hunk 1 above)
Comment (ewels) [reply main.nf]: I think the only way to handle this is to have a second conditional join that adds on the alignment logs if we're not merging. This isn't very tidy, but I can't think of another way to handle both scenarios. Except just skipping the `bismark2report` step if we're merging.

File: main.nf [hunk 3]
```diff
@@ -391,8 +395,10 @@ if( params.aligner == 'bismark' ){
         file wherearemyfiles from ch_wherearemyfiles_for_bismark_align.collect()
 
         output:
-        set val(name), file("*.bam") into ch_bam_for_bismark_deduplicate, ch_bam_for_bismark_summary, ch_bam_for_preseq
-        set val(name), file("*report.txt") into ch_bismark_align_log_for_bismark_report, ch_bismark_align_log_for_bismark_summary, ch_bismark_align_log_for_multiqc
+        file("*.bam") into ch_bam_for_merge
```
Comment (ewels) [inline main.nf]: I don't think you need any of these changes. Just switching `ch_bam_for_bismark_deduplicate` for `ch_bam_for_merge` should be sufficient.

File: main.nf [hunk 4]
```diff
@@ -364,7 +368,7 @@ if( params.notrim ){
             """
         } else {
             """
-            trim_galore --paired --fastqc --gzip $rrbs $c_r1 $c_r2 $tpc_r1 $tpc_r2 $reads
+           trim_galore --paired --fastqc --gzip $rrbs $c_r1 $c_r2 $tpc_r1 $tpc_r2 $reads
```
Comment (ewels) [inline main.nf]: Please revert the whitespace changes

File: main.nf [hunk 5]
```diff
@@ -350,7 +354,7 @@ if( params.notrim ){
         set val(name), file('*fq.gz') into ch_trimmed_reads_for_alignment
         file "*trimming_report.txt" into ch_trim_galore_results_for_multiqc
         file "*_fastqc.{zip,html}"
-        file "where_are_my_files.txt"
+        //file "where_are_my_files.txt"
```
Comment (ewels) [inline main.nf]: Please revert

File: main.nf [hunk 6]
```diff
@@ -314,16 +315,19 @@ process fastqc {
 
     input:
     set val(name), file(reads) from ch_read_files_for_fastqc
+    //file wherearemyfiles from ch_wherearemyfiles_for_fastqc
```
Comment (ewels) [inline main.nf]: Please revert

File: main.nf [hunk 7]
```diff
@@ -30,8 +31,8 @@ params.fasta_index = params.genome ? params.genomes[ params.genome ].fasta_index
 assert params.aligner == 'bwameth' || params.aligner == 'bismark' : "Invalid aligner option: ${params.aligner}. Valid options: 'bismark', 'bwameth'"
 
 Channel
-    .fromPath("$baseDir/assets/where_are_my_files.txt", checkIfExists: true)
-    .into { ch_wherearemyfiles_for_trimgalore; ch_wherearemyfiles_for_alignment }
+    .fromPath("$baseDir/assets/where_are_my_files.txt", checkIfExists: false)
+    .into {ch_wherearemyfiles_for_fastqc; ch_wherearemyfiles_for_trimgalore; ch_wherearemyfiles_for_alignment }
```
Comment (ewels) [inline main.nf]: I'm not sure why you need a `ch_wherearemyfiles_for_fastqc` for FastQC? These are placeholders for steps where we don't save the output files to `.results` by default. We always save FastQC reports, so no need for it there.

File: main.nf [hunk 8]
```diff
@@ -30,8 +31,8 @@ params.fasta_index = params.genome ? params.genomes[ params.genome ].fasta_index
 assert params.aligner == 'bwameth' || params.aligner == 'bismark' : "Invalid aligner option: ${params.aligner}. Valid options: 'bismark', 'bwameth'"
 
 Channel
-    .fromPath("$baseDir/assets/where_are_my_files.txt", checkIfExists: true)
-    .into { ch_wherearemyfiles_for_trimgalore; ch_wherearemyfiles_for_alignment }
+    .fromPath("$baseDir/assets/where_are_my_files.txt", checkIfExists: false)
```
Comment (ewels) [inline main.nf]: Why `checkIfExists: false`?

**Comment by MaxUlysse** [comment]
@sb43 Any update on this PR?

**Comment by ewels** [comment]
Closing due to lack of response.

Merging of BAMs will be added as a feature soon(ish) anyway, when the pipeline is ported to DSL2 and switched to use a sample sheet to handle input files.

**Pull request closed without merging**